llvm/lib/CodeGen/RemoveLoadsIntoFakeUses.cpp

```diff
@@ -0,0 +1,170 @@
+//===---- RemoveLoadsIntoFakeUses.cpp - Remove loads with no real uses ----===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// The FAKE_USE instruction is used to preserve certain values through
+/// optimizations for the sake of debugging. This may result in spilled values
+/// being loaded into registers that are only used by FAKE_USEs; this is not
+/// necessary for debugging purposes, because at that point the value must be on
+/// the stack and hence available for debugging. Therefore, this pass removes
+/// loads that are only used by FAKE_USEs.
+///
+/// This pass should run as late as possible, to ensure that we don't
+/// inadvertently shorten stack lifetimes by removing these loads, since the
+/// FAKE_USEs will also no longer be in effect.
+///
+//===----------------------------------------------------------------------===//
+
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/ADT/Statistic.h"
+#include "llvm/CodeGen/LiveRegUnits.h"
+#include "llvm/CodeGen/MachineFunction.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineRegisterInfo.h"
+#include "llvm/CodeGen/TargetSubtargetInfo.h"
+#include "llvm/IR/Function.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Support/Debug.h"
+
+using namespace llvm;
+
+#define DEBUG_TYPE "remove-loads-into-fake-uses"
+
+STATISTIC(NumLoadsDeleted, "Number of dead load instructions deleted");
+STATISTIC(NumFakeUsesDeleted, "Number of FAKE_USE instructions deleted");
+
+class RemoveLoadsIntoFakeUses : public MachineFunctionPass {
+public:
+  static char ID;
+
+  RemoveLoadsIntoFakeUses() : MachineFunctionPass(ID) {
+    initializeRemoveLoadsIntoFakeUsesPass(*PassRegistry::getPassRegistry());
+  }
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.setPreservesCFG();
+    MachineFunctionPass::getAnalysisUsage(AU);
+  }
+
+  StringRef getPassName() const override {
+    return "Remove Loads Into Fake Uses";
+  }
+
+  bool runOnMachineFunction(MachineFunction &MF) override;
+};
+
+char RemoveLoadsIntoFakeUses::ID = 0;
+char &llvm::RemoveLoadsIntoFakeUsesID = RemoveLoadsIntoFakeUses::ID;
+
+INITIALIZE_PASS_BEGIN(RemoveLoadsIntoFakeUses, DEBUG_TYPE,
+                      "Remove Loads Into Fake Uses", false, false)
+INITIALIZE_PASS_END(RemoveLoadsIntoFakeUses, DEBUG_TYPE,
+                    "Remove Loads Into Fake Uses", false, false)
+
+bool RemoveLoadsIntoFakeUses::runOnMachineFunction(MachineFunction &MF) {
```

Comment:
I wonder if we can just do this in ExpandPostRAPseudos and save having yet another pass.

This would also bring the side effect where I think we should have ExpandPostRAPseudos run in reverse and lazily maintain LiveRegUnits

Reply:
I'm fine to move it into another pass, and patch up that pass as I go through it - just a quick question though: Might this step too far outside the scope of what `ExpandPostRAPseudos` is expected to do? If so, would that justify changing the name slightly, or do we prefer to keep those fixed over time?

And also a minor unrelated, I noticed that passes may be added to `TargetPassConfig::addMachinePasses` or to `CodeGenPassBuilder<Derived>::addMachinePasses`, and most (but not all) passes are in both - is there an important distinction between them for determining where a new pass should go?

Reply:
I do believe we have an elide-reload-of-kill optimization but a 1 minute scan didn't find it, wherever that happens would be the best place.
TargetPassConfig::addMachinePasses = OldPM CodeGenPassBuilder = NewPM and incomplete

Reply:
Might be `MachineLateInstrsCleanup`? I see that that relies on `isKill()` for liveness checks as well, but might otherwise make sense as a location. It's hard to tell whether there are any optimizations afterwards that might cause an issue; it'd probably be faster to implement and test it than scan through the source of all the applicable passes.

Reply:
It's not MachineLateInstrsCleanup  (which is also running in the wrong direction). It's looking for killed register operands, not KILL instructions